connectors/byoc.py

```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
-            "status": e2str(self.status),
-            "error": None,
-            "deleted_document_count": 0,
-            "indexed_document_count": 0,
-            "created_at": iso_utc(self.created_at),
-            "completed_at": None,
-        }
-        resp = await self.client.index(index=JOBS_INDEX, document=job_def)
-        self.job_id = resp["_id"]
-        return self.job_id
+    @property
+    def configuration(self):
+        return DataSourceConfiguration(self.get("connector", "configuration"))
 
-    async def done(self, indexed_count=0, deleted_count=0, exception=None):
-        self.completed_at = datetime.now(timezone.utc)
+    @property
+    def filtering(self):
+        return Filter(self.get("connector", "filtering", default={}))
 
-        job_def = {
-            "deleted_document_count": deleted_count,
-            "indexed_document_count": indexed_count,
-            "completed_at": iso_utc(self.completed_at),
-        }
+    @property
+    def pipeline(self):
+        return Pipeline(self.get("connector", "pipeline"))
 
-        if exception is None:
-            self.status = JobStatus.COMPLETED
-            job_def["error"] = None
-        else:
-            self.status = JobStatus.ERROR
-            job_def["error"] = str(exception)
+    @property
+    def terminated(self):
+        return self.status in (JobStatus.ERROR, JobStatus.COMPLETED, JobStatus.CANCELED)
 
-        job_def["status"] = e2str(self.status)
+    @property
+    def indexed_document_count(self):
+        return self.get("indexed_document_count", default=0)
 
-        return await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    @property
+    def indexed_document_volume(self):
+        return self.get("indexed_document_volume", default=0)
 
-    async def suspend(self):
-        self.status = JobStatus.SUSPENDED
-        job_def = {"status": e2str(self.status)}
+    @property
+    def deleted_document_count(self):
+        return self.get("deleted_document_count", default=0)
 
-        await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    @property
+    def total_document_count(self):
+        return self.get("total_document_count", default=0)
 
-    @classmethod
-    def transform_filtering(cls, filtering):
-        # deepcopy to not change the reference resulting in changing .elastic-connectors filtering
-        filtering = (
-            {"advanced_snippet": {}, "rules": []}
-            if (filtering is None or len(filtering) == 0)
-            else deepcopy(filtering)
+    async def validate_filtering(self, validator):
+        validation_result = await validator.validate_filtering(self.filtering)
+
+        if validation_result.state != FilteringValidationState.VALID:
+            raise InvalidFilteringError(
+                f"Filtering in state {validation_result.state}. Expected: {FilteringValidationState.VALID}."
+            )
+        if len(validation_result.errors):
```

Comment:
Hi @timgrein , is there any scenario where the state is `VALID` but `errors` is not empty?

Reply:
No scenario from our side, but I want to be as robust as possible against connector implementations outside of our control. We could use an invariant (similar to "Design by Contract") and add an `assert` here, which simply signals a connector developer, that the provided filtering implementation is semantically wrong (or we are not as strict and log a warning or an error).

Reply:
I kind of think this is part of our protocol, although it's not explicitly stated, it's just a convention.

In another PR, the draft rules will overwrite active rules, if the state is `VALID` after validation. I don't think we need to make it complicated.